src/Features/Core/Portable/CodeRefactorings/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,157 @@
+            if (switchExpression == null)
+            {
+                switchExpression = expression;
+                return true;
+            }
+
+            return AreEquivalentCore(expression, switchExpression);
+        }
+
+        private static bool IsConstant(TExpressionSyntax node, SemanticModel semanticModel)
+            => semanticModel.GetConstantValue(node).HasValue;
+
+        protected static bool TryDetermineConstant(
+            TExpressionSyntax expression1,
+            TExpressionSyntax expression2,
+            SemanticModel semanticModel,
+            out TExpressionSyntax constant,
+            out TExpressionSyntax expression)
+        {
+            (constant, expression) =
+                    IsConstant(expression1, semanticModel) ? (expression1, expression2) :
+                    IsConstant(expression2, semanticModel) ? (expression2, expression1) :
+                    default((TExpressionSyntax, TExpressionSyntax));
+
+            return constant != null;
+        }
+
+        protected abstract SyntaxToken GetIfKeyword(TIfStatementSyntax ifStatement);
+        protected abstract bool AreEquivalentCore(TExpressionSyntax expression, TExpressionSyntax switchExpression);
+        protected abstract bool CanConvertIfToSwitch(TIfStatementSyntax ifStatement, SemanticModel semanticModel);
+        protected abstract IEnumerable<TExpressionSyntax>
+            GetLogicalOrExpressionOperands(TExpressionSyntax syntaxNode);
+
+        protected abstract IEnumerable<(TStatementSyntax, TExpressionSyntax)>
+            GetIfElseStatementChain(TIfStatementSyntax ifStatement);
+
+        protected abstract TPattern CreatePatternFromExpression(
+            TExpressionSyntax operand, SemanticModel semanticModel, ref TExpressionSyntax switchExpression);
+
+        protected abstract SyntaxNode CreateSwitchStatement(
+            TStatementSyntax switchDefaultBody,
+            TExpressionSyntax switchExpression,
+            SemanticModel semanticModel,
+            List<(List<TPattern> patterns, TStatementSyntax body)> sections);
+
+        private Task<Document> UpdateDocumentAsync(
+            SyntaxNode root,
+            Document document,
+            TIfStatementSyntax ifStatement,
+            TStatementSyntax switchDefaultBody,
+            TExpressionSyntax switchExpression,
+            SemanticModel semanticModel,
+            List<(List<TPattern>, TStatementSyntax)> sections)
+        {
+            var @switch = CreateSwitchStatement(switchDefaultBody, switchExpression, semanticModel, sections);
+            var newRoot = root.ReplaceNode(ifStatement, @switch);
+            return Task.FromResult(document.WithSyntaxRoot(newRoot));
+        }
+
+        protected sealed class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(FeaturesResources.Convert_if_to_switch, createChangedDocument)
```

Comment:
instead of directly referencing the resource here.  just take in the 'title' as an argument.  Then have the caller of this constructor use an abstract property to get the title.  Then you can override the title for VB vs C#.